src/main/java/wolox/training/controllers/UserController.java

```diff
@@ -0,0 +1,84 @@
+package wolox.training.controllers;
+
+import javax.validation.constraints.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.server.ResponseStatusException;
+import wolox.training.exceptions.UserIdMismatchException;
+import wolox.training.exceptions.UserNotFoundException;
+import wolox.training.models.Book;
+import wolox.training.models.User;
+import wolox.training.repositories.UserRepository;
+
+@RestController
+@RequestMapping("/api/users")
+public class UserController {
+
+  @Autowired
+  private UserRepository userRepository;
+
+  @GetMapping
+  public Iterable findAll() {
+    return userRepository.findAll();
+  }
+
+  @GetMapping("/{id}")
+  public User findOne(@PathVariable Long id) {
+    try {
+      return userRepository.findById(id).orElseThrow(
+          () -> new UserNotFoundException("User not found for show")
+      );
+    } catch (UserNotFoundException ex) {
+      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User Not Found", ex);
+    }
+  }
+
+  @PostMapping
+  @ResponseStatus(HttpStatus.CREATED)
+  public User create(@RequestBody User user) {
+    return userRepository.save(user);
+  }
+
+  @DeleteMapping("/{id}")
+  public void delete(@PathVariable Long id) {
+    try {
+      userRepository.findById(id).orElseThrow(
+          () -> new UserNotFoundException("User not found when deleting"));
+      userRepository.deleteById(id);
+    } catch (UserNotFoundException ex) {
+      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User Not Found", ex);
+    }
+  }
+
+  @PutMapping("/{id}")
+  public User updateUser(@NotNull @RequestBody User user,
+      @PathVariable Long id) {
+    User userFromParams = findUserByIdFromParams(user, id);
+    return userRepository.save(userFromParams);
```

Comment:
Actually, you should do the save on the user that came on the RequestBody, given that's the one with the updated fields. It's perfect that you check that it's id is correct and existant, but the save is taking effect on the user you got from the DB.

Reply:
That makes a lot of sense, yes